colbase: introduce new package

yuzefovich opened the pull request:
**colbase: create new package**

This commit introduces `colbase` package which currently contains the
following (that are extracted from `colexec` package):
- `allocator.go`
- `random_testutils.go`
- `Operator` interface
- `CopyBatch` method.

The reason for extracting these things is that they are used by multiple
col* packages, so I think it's a good hygiene to separate them out.

Release note: None

**vecerror: rename and move execerror package**

This commit first renames `execerror` package to `vecerror` and then
moves it inside of `colbase` directory. It also changes the panic
matching so that now we catch all panics coming from a package that has
`pkg/sql/col` in its path (this will make sure that we don't forget to
register newly added packages that use panic-catching mechanism of the
vectorized engine with the panic-catcher). Additionally, this commit
renames the methods of `execerror` package.

Release note: None

**colbase: move typeconv package and testutils file**

This commit moves `colexec/typeconv` package into `colbase` as well as
`colexec/testutils.go` file.

Release note: None

**vecerror: remove NonVectorizedPanic**

This commit removes `vecerror.NonVectorizedPanic` in favor of
`vecerror.ExpectedError`. The guidance on whether `InternalError` or
`ExpectedError` method should be used has been updated: the distinction
is whether the vectorized engine ends up in an unexpected - invalid -
state (for example, we expect that the vectorized engine might be
performing division by zero when evaluating a binary expression).

Release note: None

yuzefovich committed:
colbase: create new package

This commit introduces `colbase` package which currently contains the
following (that are extracted from `colexec` package):
- `allocator.go`
- `random_testutils.go`
- `Operator` interface
- `CopyBatch` method.
The reason for extracting these things is that they are used by multiple
col* packages, so I think it's a good hygiene to separate them out.

Release note: None

yuzefovich committed:
vecerror: rename and move execerror package

This commit first renames `execerror` package to `vecerror` and then
moves it inside of `colbase` directory. It also changes the panic
matching so that now we catch all panics coming from a package that has
`pkg/sql/col` in its path (this will make sure that we don't forget to
register newly added packages that use panic-catching mechanism of the
vectorized engine with the panic-catcher). Additionally, this commit
renames the methods of `execerror` package.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47547)
<!-- Reviewable:end -->

yuzefovich committed:
colbase: move typeconv package and testutils file

This commit moves `colexec/typeconv` package into `colbase` as well as
`colexec/testutils.go` file.

Release note: None

yuzefovich commented:
I'm curious to hear other ideas for the naming of this new package, maybe `colexecbase` or `vecbase` would be better?

yuzefovich committed:
vecerror: remove NonVectorizedPanic

This commit removes `vecerror.NonVectorizedPanic` in favor of
`vecerror.ExpectedError`. The guidance on whether `InternalError` or
`ExpectedError` method should be used has been updated: the distinction
is whether the vectorized engine ends up in an unexpected - invalid
- state (for example, we expect that the vectorized engine might be
performing division by zero when evaluating a binary expression).

Release note: None

asubiotto reviewed (changes requested):
**Commit 1**:
The `colbase` package seems like it's just got a lot of things that don't belong in `colexec`. For example, the testing utilities seem like they can be moved to `coldata` since they're specifically about generating random data and you're probably already importing that if working with batches. Alternatively it should go into some `coldatatestutils` package or something. I think that should probably also include `CopyBatch` since it's used only in testing code and is supposed to be more of a testing utility anyway.

Regarding `Operator`, could you give an example of how it's beneficial to move it to a separate package? I don't doubt it because we do the same thing with the `Processor` interface in `execinfra`, but just want to wrap my head around it. Not sure what we should name this package. Maybe `colexecbase` is fine cc @jordanlewis 

The allocator seems like the thing we want to move out of `colexec` the most. Does it make sense to move it to `coldata` or is there some obstacle? Alternatively I think we should move it to its own package.

**Commit 2**:
I don't like the introduction of our first package (as far as I know) with the `vec` prefix if everything else is `col`. Maybe we should name more things `vec` (e.g.`vecexec`) but looks like `col` is here to stay. Should this be `colexecerror` instead?

**Commit 3**:
Can this be squashed into commit 1? The same `testutils` opinion applies here. Also why not move `typeconv` into `coltypes`?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47547#-:-M56uA_CBGbxv4_y9UlM:bvz6xoy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Hm, it's going to be very painful to adjust the commits in this PR since I have another big commit on top of these 4 in another PR. I propose that I will close this PR, will squash four commits into one in another PR and then address the comments in a new commit.
